service/controller/resource/instance/internal/vmsscheck/vmss.go

```diff
@@ -48,6 +73,16 @@ func InstancesAreRunning(ctx context.Context, logger micrologger.Logger, rg stri
 		}
 	}
 
+	if allSucceeded {
+		// All instances are succeeded, let's check the VMSS rate is safe.
+		// If current remaining API calls are less than the desired threshold, we don't proceed.
```

Comment:
I'm not very fond of this. This gets easily forgotten and hidden and is not obvious from function name nor godoc comment.

I'd propose to introduce an error & matcher and then handle the error correspondingly. To me it also seems that it might make sense to do the same thing for few other functions such as `r.getInstancesCount()`.

That way the situation is lot easier to understand when following the higher level control flow and it doesn't come as a surprise that if VMSS rate limit quotas are low that suddenly the state transitioning doesn't take place even though all instances are clearly `Succeeded` in portal.

Reply:
I agree with the custom error idea, but I don't get why we should do that for the `r.getInstancesCount` function. That is a read only function that does not cause the VMSS rate to increase at all.